packages/vite/src/node/plugins/html.ts

```diff
@@ -331,111 +383,109 @@ export function buildHtmlPlugin(config: ResolvedConfig): Plugin {
-            for (const p of node.props) {
-              if (
-                p.type === NodeTypes.ATTRIBUTE &&
-                p.value &&
-                assetAttrs.includes(p.name)
-              ) {
+            for (const p of node.attrs) {
+              if (p.value && assetAttrs.includes(p.name)) {
+                const attrSourceCodeLocation =
+                  node.sourceCodeLocation!.attrs![p.name]
                 // assetsUrl may be encodeURI
-                const url = decodeURI(p.value.content)
+                const url = decodeURI(p.value)
                 if (!isExcludedUrl(url)) {
                   if (
-                    node.tag === 'link' &&
+                    node.nodeName === 'link' &&
                     isCSSRequest(url) &&
                     // should not be converted if following attributes are present (#6748)
-                    !node.props.some(
+                    !node.attrs.some(
                       (p) => p.name === 'media' || p.name === 'disabled'
                     )
                   ) {
                     // CSS references, convert to import
                     const importExpression = `\nimport ${JSON.stringify(url)}`
                     styleUrls.push({
                       url,
-                      start: node.loc.start.offset,
-                      end: node.loc.end.offset
+                      start: node.sourceCodeLocation!.startOffset,
+                      end: node.sourceCodeLocation!.endOffset
                     })
                     js += importExpression
                   } else {
-                    assetUrls.push(p)
+                    assetUrls.push({
+                      attr: p,
+                      sourceCodeLocation: attrSourceCodeLocation
+                    })
                   }
                 } else if (checkPublicFile(url, config)) {
-                  s.overwrite(
-                    p.value.loc.start.offset,
-                    p.value.loc.end.offset,
-                    `"${toOutputPublicFilePath(url)}"`,
-                    { contentOnly: true }
+                  overwriteAttrValue(
+                    s,
+                    attrSourceCodeLocation,
+                    toOutputPublicFilePath(url)
                   )
                 }
               }
             }
           }
           // <tag style="... url(...) ..."></tag>
           // extract inline styles as virtual css and add class attribute to tag for selecting
-          const inlineStyle = node.props.find(
-            (prop) =>
-              prop.name === 'style' &&
-              prop.type === NodeTypes.ATTRIBUTE &&
-              prop.value &&
-              prop.value.content.includes('url(') // only url(...) in css need to emit file
-          ) as AttributeNode
+          const inlineStyle = node.attrs.find(
+            (prop) => prop.name === 'style' && prop.value.includes('url(') // only url(...) in css need to emit file
+          )
```

Comment:
```js
          const styleAttr = node.attrs.find((prop) => prop.name === 'style')
          
          if (!styleAttr.value.includes('url(')) // only url(...) in css need to emit file
```
This change is not necessary but I think this is more correct when there is duplicate style attributes.